File: app/src/main/java/org/oppia/app/backend/ModifyJsonInterceptor.java
package org.oppia.app.backend;

import android.util.Log;

import java.io.IOException;

import okhttp3.Interceptor;
import okhttp3.MediaType;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;

public class ModifyJsonInterceptor implements Interceptor {

  //TODO: Transfer this XSSI_PREFIX to a constant file
  // which is responsible for networking too.
  private final static String XSSI_PREFIX = ")]}\'\n";
  private static final String TAG = ModifyJsonInterceptor.class.getSimpleName();

  @Override
  public Response intercept(Chain chain) throws IOException {

    Request request = chain.request();
    Response response = chain.proceed(request);

    if (response.code() == 200) {
      if (response.body() != null) {
        String rawJson = response.body().string();

        Log.d(TAG, "intercept: rawJson: " + rawJson);

        if (rawJson.startsWith(XSSI_PREFIX)) {
          rawJson = rawJson.substring(rawJson.indexOf('\n') + 1);
        }

        Log.d(TAG, "intercept: final: " + rawJson);

        MediaType contentType = response.body().contentType();
        ResponseBody body = ResponseBody.create(contentType, rawJson);
        return response.newBuilder().body(body).build();
      }
    } else if (response.code() == 403) {
      //TODO: Manage other network errors here
    }

    return response;

  }

}

File: data/src/main/java/org/oppia/data/backends/gae/ModifyJsonInterceptor.kt
package org.oppia.data.backends.gae

import okhttp3.*

import java.io.IOException

//TODO: Transfer this XSSI_PREFIX to a constant file
// which is responsible for networking too.
private const val XSSI_PREFIX = ")]}\'\n"

class ModifyJsonInterceptor : Interceptor {

  @Throws(IOException::class)
  override fun intercept(chain: Interceptor.Chain): Response {
    val request = chain.request()
    val response = chain.proceed(request)

    if (response.code() == 200) {
      if (response.body() != null) {
        var rawJson = response.body()!!.string()
        if (rawJson.startsWith(XSSI_PREFIX)) {
          rawJson = rawJson.substring(rawJson.indexOf('\n') + 1)
        }
        val contentType = response.body()!!.contentType()
        val body = ResponseBody.create(contentType, rawJson)
        return response.newBuilder().body(body).build()
      }
    } else if (response.code() == 403) {
      //TODO: Manage other network errors here
    }
    return response
  }

}
